c/secp256k1_blake160_multisig_all.c

```diff
@@ -0,0 +1,226 @@
+#include "blake2b.h"
+#include "ckb_syscalls.h"
+#include "protocol_reader.h"
+#include "secp256k1_helper.h"
+
+#define ERROR_ARGUMENTS_LEN -1
+#define ERROR_ENCODING -2
+#define ERROR_SYSCALL -3
+#define ERROR_SECP_RECOVER_PUBKEY -11
+#define ERROR_SECP_PARSE_SIGNATURE -12
+#define ERROR_SECP_SERIALIZE_PUBKEY -13
+#define ERROR_SCRIPT_TOO_LONG -21
+#define ERROR_WITNESS_TOO_LONG -22
+#define ERROR_WITNESS_TOO_SHORT -23
+#define ERROR_INVALID_PUBKEYS_CNT -24
+#define ERROR_INVALID_THRESHOLD -25
+#define ERROR_INVALID_REQUIRE_FIRST_N -26
+#define ERROR_MULTSIG_SCRIPT_HASH -31
+#define ERROR_VERIFICATION -32
+
+#define BLAKE2B_BLOCK_SIZE 32
+#define BLAKE160_SIZE 20
+#define PUBKEY_SIZE 33
+#define TEMP_SIZE 1024
+#define RECID_INDEX 64
+/* 32 KB */
+#define WITNESS_SIZE 32768
+#define SCRIPT_SIZE 32768
+#define SIGNATURE_SIZE 65
+#define FLAGS_SIZE 4
+
+/*
+ * Arguments:
+ * multisig script blake160 hash, 20 bytes.
+ *
+ * Witness:
+ * multisig_script | Signature1 | signature2 | ...
+ * multisig_script: S | R | M | N | Pubkey1 | Pubkey2 | ...
+ *
+ * +------------+----------------------------------+-------+
+ * |            |           Description            | Bytes |
+ * +------------+----------------------------------+-------+
+ * | S          | reserved for future use          |     1 |
+ * | R          | first nth public keys must match |     1 |
+ * | M          | threshold                        |     1 |
+ * | N          | total public keys                |     1 |
+ * | PubkeyN    | compressed pubkey                |    33 |
+ * | SignatureN | recoverable signature            |    65 |
+ * +------------+----------------------------------+-------+
+ *
+ */
+
+int main() {
+  int ret;
+  volatile uint64_t len = 0;
+  unsigned char temp[TEMP_SIZE];
+
+  /* Load args */
+  unsigned char script[SCRIPT_SIZE];
+  len = SCRIPT_SIZE;
+  ret = ckb_load_script(script, &len, 0);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+  if (len > SCRIPT_SIZE) {
+    return ERROR_SCRIPT_TOO_LONG;
+  }
+  mol_pos_t script_pos;
+  script_pos.ptr = (const uint8_t*)script;
+  script_pos.size = len;
+
+  mol_read_res_t args_res = mol_cut(&script_pos, MOL_Script_args());
+  if (args_res.code != 0) {
+    return ERROR_ENCODING;
+  }
+  mol_read_res_t args_bytes_res = mol_cut_bytes(&args_res.pos);
+  if (args_bytes_res.code != 0) {
+    return ERROR_ENCODING;
+  } else if (args_bytes_res.pos.size != BLAKE160_SIZE) {
+    return ERROR_ARGUMENTS_LEN;
+  }
+
+  /* Load witness of first input */
+  unsigned char witness[WITNESS_SIZE];
+  len = WITNESS_SIZE;
+  ret = ckb_load_witness(witness, &len, 0, 0, CKB_SOURCE_GROUP_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+  if (len > WITNESS_SIZE) {
+    return ERROR_WITNESS_TOO_LONG;
+  }
+  if (len < FLAGS_SIZE) {
+    return ERROR_WITNESS_TOO_SHORT;
+  }
+
+  /* Get flags */
```

Comment:
Depending on the personal view this might or might not be an issue, but I feel like we are re-inventing a new simple serialization format here to avoid using molecule. While I do agree using a simple format here would save some effort here alone, using molecule might be a choice to think since it allows saving efforts from a global perspective. E.g.: client libraries will have to implement the same format here as well to use the script, and what if the scripts get updated with more features in the future?

Reply:
Unrelated to this PR. This was introduced in https://github.com/nervosnetwork/ckb-system-scripts/pull/48